Method Reduce::exactlyN() added

Smoren opened the pull request:
1. Method `Reduce::exactlyN()` added
2. Tests added (coverage 100%).
3. README updated with new method description.

Smoren committed:
Method `Reduce::exactlyN` added (using `Multi::zipStrict()`).

Smoren committed:
`Reduce::exactlyN()`: applied `Multi::zipEqual()` and `\LengthException`.

Smoren committed:
`Reduce::exactlyN()`: old implementation removed.

markrogoyski commented on src/Reduce.php:
```diff
@@ -175,4 +175,32 @@ public static function isReversed(iterable $data): bool
+     *
+     * @return bool
+     */
+    public static function exactlyN(iterable $data, iterable $n, iterable ...$extra): bool
+    {
+        $iterables = [$data, $n, ...$extra];
```
Question: Are the three different parameters needed? It seems like you can just have the method declaration be a single variadic parameter like this:
```php
public static function exactlyN(iterable ...$iterables): bool
```

And then pass that to zip and not need the extra variable assignment here.

Smoren replied:
Done.

markrogoyski commented on src/Reduce.php:
```diff
@@ -175,4 +175,32 @@ public static function isReversed(iterable $data): bool
+     *
+     * @return bool
+     */
+    public static function exactlyN(iterable $data, iterable $n, iterable ...$extra): bool
```
Suggestion: How about naming this `same` or something. `exactlyN` generally has a different meaning, as in, the iterable/collection has exactly n elements (for example, exactly 10 elements). It seems like the function you are proposing is "do the iterables produce the same collection of elements?" Therefore, `same` seems to be a better fit here.

Smoren replied:
Done.

markrogoyski commented:
Hi @Smoren,

Thank you for your continued interest in IterTools and for the pull request. Thank you for adding many unit test cases.

I left a couple comments inline. Also, a general comment about unit tests: I recommend not spending time to line up all the test cases in neat columns. The reason for this is that if you add a single test case that is longer than any previous one, you will again have to spend time rearranging them again into a new format. I recommend just putting each argument on it's own line for simplicity so everything is formatted the same way.

Smoren committed:
1. `Reduce::exavtlyN()` renamed to `Reduce::same`.
2. Tests updated.
3. README updated.

Smoren commented:
Thank you for your comments!

All your suggestions are applied.

1. `Reduce::exavtlyN()` renamed to `Reduce::same()`.
2. Tests updated.
3. README updated.

Smoren committed:
`Reduce::same()`: tests fixed.

Smoren committed:
1. Method `Reduce::sameCount()` added.
2. Tests added for the new method (coverage 100%).
3. README updated with the description of the new method.

Smoren commented:
Hi @markrogoyski

I've added a new method `sameCount`.

1. Method `Reduce::sameCount()` added.
2. Tests added for the new method (coverage 100%).
3. README updated with the description of the new method.

markrogoyski commented:
Hi @Smoren,

I refactored some of the new code in the `develop` branch. I've made a distinction between `Reduce` and a new class I've named `Summary`. The `isSorted` and `isReversed` are conceptually different from the other functions in `Reduce`, as they are not reducing the collection to a single value, but rather summarizing some aspect of the collection (or collections) by answering a question, like is the iterable sorted? The answers to the questions are true or false.

I think this caused some git conflicts in your pull request. Apologies for that.

I think the new functions in your pull request would go in the new `Summary` class, as they are answering questions about the collections, rather than reducing to a value. Are they they same, and do they have the same number of items, are summarizing queries that have a yes/no (true/false) answer.

Let me know if you have any questions about the new namespace for the functions? And sorry for the extra work to move them. Please pull in the latest changes from `develop` and it should be clear what the changes are.

Thanks again for the pull request.
Mark

markrogoyski commented on src/Reduce.php:
```diff
@@ -175,4 +175,55 @@ public static function isReversed(iterable $data): bool
+     *
+     * @return bool
+     */
+    public static function sameCount(iterable ...$iterables): bool
+    {
+        $count = null;
+        foreach ($iterables as $iterable) {
+            $currentCount = self::toCount($iterable);
+
+            if ($count !== null && $count !== $currentCount) {
+                return false;
+            }
+
+            $count = $currentCount;
+        }
+
+        return true;
```
I wonder if this could be simplified by using some functional programming constructs.

For example, would something like this work and be easier to follow:
```php
if (\count($iterables) <= 1) {
    return true;
}

$counts = \array_map('self::toCount', $iterables);
return \count(\array_unique($counts)) === 1;
```

Smoren replied:
Done.

Smoren committed:
`Reduce::sameCount()`: method's body changed.

Smoren committed:
1. Methods `same` and `sameCount` moved to namespace `Summary`.
2. Tests moved to directory `Summary`.
3. README updated.

Smoren commented:
Hi @markrogoyski 

All done :)

Sorry for making you wait.

1. All conflicts are resolved during the rebasing.
2. Methods `same` and `sameCount` are moved to namespace `Summary`.
3. Tests are moved to directory `Summary`.
4. README updated.

Smoren committed:
README fix of the descriptions order (Is Sorted, Is reversed).

markrogoyski commented:
Hi @Smoren,

Thank you again for the pull request. I've merged it into develop to go out with the next release.

Smoren commented:
Thank you Mark!

markrogoyski commented:
Hi @Smoren,

The new functionality in this PR, as well as your other PRs, has been released in the latest release [v1.1.0](https://github.com/markrogoyski/itertools-php/releases/tag/v1.1.0). Thanks again!

Smoren commented:
Thank you Mark! Release is great and long-awaited news!
